comet-java-client/src/main/java/ml/comet/experiment/impl/ArtifactImpl.java

```diff
@@ -0,0 +1,216 @@
+package ml.comet.experiment.impl;
+
+import com.vdurmont.semver4j.Semver;
+import lombok.Getter;
+import lombok.NonNull;
+import ml.comet.experiment.artifact.Artifact;
+import ml.comet.experiment.artifact.ArtifactBuilder;
+import ml.comet.experiment.artifact.ConflictingArtifactAssetNameException;
+import ml.comet.experiment.impl.asset.ArtifactAsset;
+import ml.comet.experiment.impl.asset.ArtifactRemoteAsset;
+import ml.comet.experiment.impl.asset.Asset;
+import ml.comet.experiment.impl.asset.RemoteAsset;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Optional.empty;
+import static ml.comet.experiment.impl.resources.LogMessages.CONFLICTING_ARTIFACT_ASSET_NAME;
+import static ml.comet.experiment.impl.resources.LogMessages.getString;
+import static ml.comet.experiment.impl.utils.AssetUtils.createAssetFromData;
+import static ml.comet.experiment.impl.utils.AssetUtils.createAssetFromFile;
+import static ml.comet.experiment.impl.utils.AssetUtils.createRemoteAsset;
+import static ml.comet.experiment.impl.utils.AssetUtils.walkFolderAssets;
+
+/**
+ * The implementation of the {@link Artifact}.
+ */
+public final class ArtifactImpl extends BaseArtifactImpl implements Artifact {
+
+    @Getter
+    private final List<Asset> assets;
+
+    private final boolean prefixWithFolderName;
+
+    ArtifactImpl(String name, String type) {
+        super(name, type);
+        this.assets = new ArrayList<>();
```

Comment:
The dictionary in Python has completely different implementation than `HashMap` or other `Map` implementations in Java.
We are not going to have duplicates because before adding to the list we explicitly check for uniqueness of the asset to be added.
Using `Map` instead of `List` here will not give us any advantage.